consensus/raft/logging.go

```diff
@@ -7,29 +7,85 @@ import (
 	logging "github.com/ipfs/go-log"
 )
 
+const (
+	debug = iota
+	info
+	warn
+	err
+)
+
 // This provides a custom logger for Raft which intercepts Raft log messages
 // and rewrites us to our own logger (for "raft" facility).
-type logForwarder struct{}
+type logForwarder struct {
+	last map[int]*lastMsg
+}
+
+type lastMsg struct {
+	msg    string
+	tipped bool
+}
 
 var raftStdLogger = log.New(&logForwarder{}, "", 0)
 var raftLogger = logging.Logger("raft")
 
 // Write forwards to our go-log logger.
 // According to https://golang.org/pkg/log/#Logger.Output
 // it is called per line.
-func (fw *logForwarder) Write(p []byte) (n int, err error) {
+func (fw *logForwarder) Write(p []byte) (n int, e error) {
 	t := strings.TrimSuffix(string(p), "\n")
+
 	switch {
 	case strings.Contains(t, "[DEBUG]"):
-		raftLogger.Debug(strings.TrimPrefix(t, "[DEBUG] raft: "))
+		if !fw.repeated(debug, t) {
+			fw.log(debug, strings.TrimPrefix(t, "[DEBUG] raft: "))
+		}
 	case strings.Contains(t, "[WARN]"):
-		raftLogger.Warning(strings.TrimPrefix(t, "[WARN]  raft: "))
+		if !fw.repeated(warn, t) {
+			fw.log(warn, strings.TrimPrefix(t, "[WARN]  raft: "))
+		}
 	case strings.Contains(t, "[ERR]"):
-		raftLogger.Error(strings.TrimPrefix(t, "[ERR] raft: "))
+		if !fw.repeated(err, t) {
+			fw.log(err, strings.TrimPrefix(t, "[ERR] raft: "))
+		}
 	case strings.Contains(t, "[INFO]"):
-		raftLogger.Info(strings.TrimPrefix(t, "[INFO] raft: "))
+		if !fw.repeated(info, t) {
+			fw.log(info, strings.TrimPrefix(t, "[INFO] raft: "))
+		}
 	default:
-		raftLogger.Debug(t)
+		fw.log(debug, t)
 	}
 	return len(p), nil
 }
+
+func (fw *logForwarder) repeated(t int, msg string) bool {
+	if fw.last == nil {
+		fw.last = make(map[int]*lastMsg)
+	}
+
+	last, ok := fw.last[t]
+	if !ok || last.msg != msg {
+		fw.last[t] = &lastMsg{msg, false}
+		return false
+	} else {
+		if !last.tipped {
+			fw.log(t, "NOTICE: The last RAFT log message repeats and will only be logged once")
```

Comment:
With logging from the Raft module we have the advantage that we are proxying messages so we can do this thing rather easily. Raft spams the logs every second when a peer is down and this bothered me (specially for debugging). This will only reduce messages a bit, since it also sends different messages. For the moment let's try it just here.